### ISSUE TITLE
docker_container and broken volume definitions

### ISSUE THREAD

**Issue by jperville** [bflad/chef-docker#428]
In the same spirit as #427, I found some differences in the `docker inspect` output when trying to declare a container which volumes, which resulted in bind mount being unusable with that container.

The test container is a container based upon the busybox:ubuntu-14.04 image which uses a bind-mount to the local filesystem (eg. `docker run ... -v /docker-images:/images/ro ...`) and at the same time bind mount from a data container using `volumes_from`.

The container is declared like this (simplified version):
```ruby
docker_container 'my-test-app' do
  repo 'busybox'
  tag 'ubuntu-14.04'
  volumes_from 'my-test-data'
  volume '/docker-images:/images:ro'
  port ['7080:80', '7022:22'] # bug #427
  command '/bin/top'
  ...
end
```

The following if the diff of `docker inspect` between:
 * the container as generated by this cookbook
 * the container recreated manually with the docker client.

TL;DR: the main difference is the `Binds` and `.Mount` entries, which are broken on the cookbook version. The diff also displays bug #427 in the `PortBindings` section and another bug of unknown consequences related to hostname/domainname parsing that will report next.

```diff
--- inspect-cookbook.json       2015-09-16 12:47:18.568194819 +0000
+++ inspect-manual.json 2015-09-16 13:01:22.084194819 +0000
@@ -63,7 +63,9 @@
     "AppArmorProfile": "",
     "ExecIDs": null,
     "HostConfig": {
-        "Binds": [],
+        "Binds": [
+            "/docker-images:/images:ro"
+        ],
         "ContainerIDFile": "",
         "LxcConf": null,
         "Memory": 67108864,
@@ -78,15 +80,15 @@
         "MemorySwappiness": null,
         "Privileged": false,
         "PortBindings": {
-            "22": [
+            "22/tcp": [
                 {
-                    "HostIp": "0.0.0.0",
+                    "HostIp": "",
                     "HostPort": "7022"
                 }
             ],
-            "80": [
+            "80/tcp": [
                 {
-                    "HostIp": "0.0.0.0",
+                    "HostIp": "",
                     "HostPort": "7080"
                 }
             ]
@@ -143,24 +145,22 @@
             "RW": true
         },
         {
-            "Name": "461673042e45a1defcb709c8971bf2eb00986553bb0e06a726d3cfadbb55d89e",
-            "Source": "/var/lib/docker/volumes/461673042e45a1defcb709c8971bf2eb00986553bb0e06a726d3cfadbb55d89e/_data",
-            "Destination": "/docker-images",
-            "Driver": "local",
-            "Mode": "",
-            "RW": true
+            "Source": "/docker-images",
+            "Destination": "/images",
+            "Mode": "ro",
+            "RW": false
         }
     ],
     "Config": {
-        "Hostname": "my-test-app.docker-ubuntu-1404",
-        "Domainname": "",
+        "Hostname": "my-test-app",
+        "Domainname": "docker-ubuntu-1404",
         "User": "",
         "AttachStdin": false,
         "AttachStdout": false,
         "AttachStderr": false,
         "ExposedPorts": {
-            "22": {},
-            "80": {}
+            "22/tcp": {},
+            "80/tcp": {}
         },
         "PublishService": "",
         "Tty": false,
@@ -184,9 +184,7 @@
             "/bin/top"
         ],
         "Image": "10.0.2.2:2000/sample-docker-app:latest",
-        "Volumes": {
-            "/docker-images:/images:ro": {}
-        },
+        "Volumes": null,
         "VolumeDriver": "",
         "WorkingDir": "",
         "Entrypoint": null,
```

**Comment by jperville** [comment]
It seems that the cookbook copies each volume definition (such as `""/docker-images:/images:ro"`) into the wrong place. Each entry ends up in `.Config.Volumes` while it should be appended to the `.Config.Bind` array.

How could I fix this while not breaking the parsing of existing volume definitions? Trying to add some integration test assertions right now and if I can make a PR today because this is blocking for me...

**Comment by someara** [comment]
This is the kind of thing they like to change between API versions (Docker
middle digit releases).

The trick is to switch on @api_version like this

https://github.com/bflad/chef-docker/blob/eed975f628a9d6fb7c290d3e7b0e352aa59f2e08/libraries/helpers_container.rb#L176

Then, over in the serverspecs, use tricks like this

https://github.com/bflad/chef-docker/blob/eed975f628a9d6fb7c290d3e7b0e352aa59f2e08/test/integration/resources-182/serverspec/assert_functioning_spec.rb#L11-L13

-s

**Comment by jperville** [comment]
So, if I understand well, volume definitions are read in .Config.Volumes but should be written to .Config.Binds (as of docker 1.8.2).

**Comment by someara** [comment]
Possibly... it could also be the case that they're simply broken above 1.6
(the version I started development against)

**Comment by jperville** [comment]
After playing in the kitchen VMs, the docker inspect is slightly different regarding volumes.

On docker 1.6.2 (default for ubuntu 14.04)
```json
{
    "HostConfig": {
        "Binds": [
            "/docker-images:/images:ro"
        ],
    },
    "Volumes": {
        "/images": "/docker-images"
    },
    "VolumesRW": {
        "/images": false
    }
}
```

On docker 1.8.2 (default for ubuntu 15.04)

```json
    "HostConfig": {
        "Binds": [
            "/docker-images:/images:ro"
        ],
    },
    "Mounts": [
        {
            "Source": "/docker-images",
            "Destination": "/images",
            "Mode": "ro",
            "RW": false
        }
    ],
```

Would writing the volume definitions into the `.Config.Bind` array using the API work in both cases?

**Comment by someara** [comment]
Only one way to find out!

**Comment by jperville** [comment]
I got it to work, the docker_container attribute which used to be called 'volumes' in version 0.37 of this cookbook is now called 'binds' (and 'volumes' is used for another purpose, maybe similar to what the VOLUME instruction does in the Dockerfile).

By doing:
```ruby
docker_container 'my-test-app' do
  repo 'busybox'
  tag 'ubuntu-14.04'
  binds '/docker-images:/images:ro' # <------ here
  command '/bin/top'
  ...
end
```

Gets me the expected bind mounts.

The use-case is documented in the README.md (search for "Bind mount local directories") but without mentioning that 'volumes' from the docker client maps to the 'binds' attribute of the docker_container resource.

Feel free to close the issue.

**Comment by someara** [comment]
Oh derp. I knew that.... that's what I get to replying to threads while
traveling =)
